NXP-27666: implement elasticSearch reindex endpoint

RSalem07 opened the pull request:
i add a new endpoint to allow the ES indexation on a given repo. once this is approved, i think we should complete the ES endpoints with a new one that give the ES indexing Status like in Jsf UI, i don't know if it's needed.  there is an Wait for indexing operation but i don't think is a good idea to start an ES indexing and then wait until the end in the same Rest resource.

nuxeojenkins commented:
[View issue in JIRA](https://jira.nuxeo.com/browse/NXP-27666)

RSalem07 commented:
for the 2nd commit, it is related to [NXP-27667](https://jira.nuxeo.com/browse/NXP-27667), as discussed i push on the same git branch because the two tickets are related. 

i add 4 endpoints (flush all / flush on repo, optimize, optimize all), for the junit tests i added the four tests to be sur that each endpoint is correctly called, but i didn't find an easy way to test the flush/optimize and i think it's not the purpose of the management API :)

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,84 @@
+     * Runs an ES indexation on the given repository.
+     *
+     * @param repositoryName the repository name for which we want to start the ES indexation
+     * @throws NuxeoException
```
not needed

RSalem07 replied:
you means the java doc ?

ataillefer replied:
`@throws NuxeoException`

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,84 @@
+        if (repository == null) {
+            throw new NuxeoException(String.format("No such repository with name: %s", repositoryName), SC_NOT_FOUND);
+        }
+
```
Extra blank line

RSalem07 replied:
already removed on the 2nd commit ; 'NXP-27667: implement elasticSearch flush and optimize endpoints'

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,126 @@
+        // Initial docs count
+        long initialDocCount = coreSession.query(GET_DOCUMENTS_QUERY).totalSize();
+
+        // Nothing indexed because of disable indexing flag
```
Nothing indexed because the index was just dropped no?
It looks like this comment makes more sens at line 88, after having created some documents with the `disableAutoIndexing` flag, then waited for the next transaction.

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,126 @@
+        // Nothing indexed because of disable indexing flag
+        assertEquals(0, ess.query(new NxQueryBuilder(coreSession).nxql(GET_DOCUMENTS_QUERY)).totalSize());
+
+        // Creates a new docs without indexing them
```
Create new documents

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,126 @@
+        createDocuments();
+
+        // Wait for an eventual indexation
+        TransactionHelper.startTransaction();
```
Shouldn't you commit the transaction first?

RSalem07 replied:
yep it's already done on the `createDocuments` method : `TransactionHelper.commitOrRollbackTransaction();`

ataillefer replied:
I think that it's not clear to have the commit separated (nested in a method) from the start.
It sounds clearer to remove the commit from the method and have commit/start together in the test.

RSalem07 replied:
i did like that because i was thinking that a create method will commit but it's not forced to create a new TX, but for this test purpose i will group them together

troger replied:
You must use `TransactionalFeature` everywhere, so use `txFeature#nextTransaction` here.

kevinleturc replied:
Please use `TransactionnalFeature#nextTransaction` just after the `createDocuments` call or as the last instruction of your method.

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Test
+    public void shouldFailRunningIndexationWhenRepositoryNotExist() {
```
NotExists

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,126 @@
+        });
+
+        coreSession.save();
+        TransactionHelper.commitOrRollbackTransaction();
```
I think that:
- you don't need `coreSession.save()` as you are commitiing the transaction
- you should start a transaction just after committing it to make sure that you always have a transaction in the test
To be checked with @efge.

kevinleturc replied:
Indeed, see my other comment. In test we should always use the transactional feature.

ataillefer commented on nuxeo-management-rest-api/src/test/resources/OSGI-INF/elasticsearch-test-contrib.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+
+<component name="org.nuxeo.elasticsearch.test.contrib">
```
If you need to define a test component here, please use a different name form the one in nuxeo-elasticsearch-core.
In fact, you might be able to deploy `org.nuxeo.elasticsearch.core.test:elasticsearch-test-contrib.xml`.

RSalem07 replied:
go for a new component, i agree ;)

ataillefer replied:
Why not try to reuse the existing one?

RSalem07 replied:
you right this file was added before the pom modification (to include the test jar dependency) and i forget about removing it and use the core one. thx

troger replied:
Agree, prefix the component name with something common to the module, such as `org.nuxeo.rest.management`.
`org.nuxeo.rest.management.elasticsearch.test`?

ataillefer commented:
This only adds an endpoint to re-index a whole repository, which is the equivalent of "Re-indexing the entire repository" in the JSF UI admin center.
I think that we should add the ability to:
- Re-index documents matching a given query, see "Re-indexing documents that match a NXQL query" in the Admin Center.
- Re-index a document and its children, see "Re-indexing a document and its children" in the Admin Center.

ataillefer commented:
I wonder if we couldn't use the `ElasticsearchBulkIndexOperation` operation to reindex the whole repository.

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,129 @@
+    @Path("{repositoryName}/reindex")
+    public void doIndexing(@PathParam("repositoryName") String repositoryName) {
+        Repository repository = getRepository(repositoryName);
+        Framework.doPrivileged(() -> {
```
We don't need the `doPrivileged` since opening a core session as system.

RSalem07 replied:
i had a doubt and i added it at the last minute,  i will remove the call to `doPrivilged`

troger replied:
Yes, use directly `CoreInstance#doPrivileged`?

kevinleturc replied:
You could also use `CoreInstance.doPrivileged` which does the `Framework.doPrivileged` and open the session based on the `repositoryName` (which will allow you to remove the `getRepository`method).

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -64,7 +65,51 @@ public void doIndexing(@PathParam("repositoryName") String repositoryName) {
+    }
+
+    /**
+     * Executes an ES optimize on all documents indexes of a given repository.
```
on the document index of a given repository (there's only one document index per repository)

RSalem07 replied:
i think i missed something, a repository will hold one index but many documents. for example we will have an User index, and many users (documents) will be indexed. it's my understanding

kevinleturc replied:
Afaik there's no user index in ES, we index the repository (one index per repository), the audit and a sequence.

In case of re-indexing, we may have a second index for the repository, representing the current indexing. Which allows us to not break service during indexation and just change the alias of the index once done.

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -64,7 +65,51 @@ public void doIndexing(@PathParam("repositoryName") String repositoryName) {
     }
 
+    /**
+     * Executes an ES flush on all documents indexes of a given repository.
```
on the document index of a given repository (there's only one document index per repository)

ataillefer commented:
Yes, a status endpoint could be nice, to be discussed with the team?

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -97,6 +97,36 @@ public void shouldRunIndexation() throws Exception {
+    @Test
+    public void shouldRunFlush() {
+        try (CloseableClientResponse response = httpClientRule.post("site/management/elasticsearch/flush", null)) {
+            assertEquals(SC_NO_CONTENT, response.getStatus());
```
Indeed, why not return something more interesting, maybe the URL of a  GET endpoint to have the status?

RSalem07 replied:
yes i was thinking to do something like that but the ES operation doesn't returns any id (at the beginning i was thinking that for each launched indexation i will have an id status), and i check the JSF admin i will find a way to returns the GET status on ES indexation.

troger commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,129 @@
+    @POST
+    @Path("{repositoryName}/reindex")
+    public void doIndexing(@PathParam("repositoryName") String repositoryName) {
+        Repository repository = getRepository(repositoryName);
```
I'm not sure we need that method, `CoreInstance.openCoreSessionSystem` will throw an exception if the repository is unknown.

RSalem07 replied:
yep this method will throw an exception of the repo is unknown, but if the `repositoryName` param is null the `openCoreSessionSystem` will take the default one and this is not what we expect, then i added the method check to validate the `repositoryName`

troger commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,129 @@
+        Framework.doPrivileged(() -> {
+            try (CloseableCoreSession coreSession = CoreInstance.openCoreSessionSystem(repository.getName())) {
+                AutomationService automationService = Framework.getService(AutomationService.class);
+                automationService.run(new OperationContext(coreSession), ElasticsearchIndexOperation.ID);
```
You could just call `ElasticSearchIndexing#reindexRepository` instead of relying on the `AutomationService`.

RSalem07 replied:
this was my first idea before starting the whole development, after a short discussion with @kevinleturc he advice me to use the Operation and i am going to use the Bulk indexing operation as proposed by @ataillefer instead of the regular one :)

troger commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,129 @@
+     * @param repositoryName the repository name.
+     * @return the {@link Repository} if exist, otherwise an exception will be thrown
+     */
+    protected Repository getRepository(String repositoryName) {
```
Not needed IMO.

troger commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,156 @@
+@Features(AutomationFeature.class)
+@Deploy("org.nuxeo.ecm.core.management")
+@Deploy("org.nuxeo.rest.management.test")
+@Deploy("org.nuxeo.ecm.automation.core")
```
Already deployed by `AutomationFeature`.

kevinleturc replied:
Not needed, it is deployed by `AutomationFeature`

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,129 @@
+    @Path("{repositoryName}/optimize")
+    public void doOptimize(@PathParam("repositoryName") String repositoryName) {
+        Repository repository = getRepository(repositoryName);
+        Framework.doPrivileged(() -> {
```
No need to use curly bracket for lambda, right?

RSalem07 replied:
sure, thx, a bad copy/past

kevinleturc commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,156 @@
+@RunWith(FeaturesRunner.class)
+@Features(AutomationFeature.class)
+@Deploy("org.nuxeo.ecm.core.management")
+@Deploy("org.nuxeo.rest.management.test")
```
Could you put this deploy at the end as it describes the current module?

RSalem07 replied:
sure :)

troger commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,156 @@
+        TransactionHelper.commitOrRollbackTransaction();
+    }
+
+    protected void waitForIndexing() throws Exception {
```
`TransactionalFeature#nextTransaction` should wait for the indexing.

kevinleturc replied:
I just had a look, and the waiter for elasticsearch indexing is in `RepositoryElasticsearchFeature` (not deployed here) and only available in 11.1.
For the work manager, we have `WorkManagerFeature` (deployed by the `CoreFeature`) which waits for it since 10.3.

ataillefer replied:
I think that this only the case if the `RepositoryElasticSearchFeature` is used.
And I'm not sure that this was backported to 10.10, see https://jira.nuxeo.com/browse/NXP-27330.

RSalem07 replied:
replace the call to `waitForIndexing` by `TransactionalFeature#nextTransaction` the `esa.refresh();` is needed (i comment this line and the test fails) what i propose is tu keep the `waitForIndexing` but i will replace it by :

```
 protected void waitForIndexing() {
        txFeature.nextTransaction();
        esa.refresh();
    }
```
wdyt

kevinleturc commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,156 @@
+    @Test
+    public void shouldFailRunningIndexationWhenRepositoryNotExist() {
+        // Launch the ES indexing
+        String path = String.format("site/management/elasticsearch/%s", "unExistingRepository");
```
You're missing the `reindex` part of the path right?

RSalem07 replied:
yep you right thx,

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,154 @@
+            coreSession.createDocument(doc);
+        });
+
+        txFeature.nextTransaction();
```
Sorry, I still find it not consistent to handle the transaction here, especially that after the call to `createDocuments(` you are calling `waitForIndexing(` that calls `txFeature.nextTransaction()`.
I think this call can just be removed here.

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,109 @@
+    @Path("{repositoryName}/reindex")
+    public void doIndexing(@PathParam("repositoryName") String repositoryName) {
+        try {
+            CoreInstance.doPrivileged(repositoryName, (coreSession) -> {
```
No bracket between the variable if number of arguments is strictly lower than 2.

RSalem07 replied:
my bad, thx

kevinleturc commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    protected void createDocuments() {
+        Stream.of("my-file-1", "my-file-2").forEach((fileName) -> {
```
Same about brackets.

RSalem07 commented:
As discussed i changed the code to allow ES indexing as below:
- Using BAF directly without automation 
- Ones an ES indexing is launched the rest resource will returns an Http 201 (Created) with the location of status
- I added the new resource to get the status: i used the commandId of the baf as an indexing Id.
- I added the endpoint to allow indexing by query (using BAF)
- For indexing Document and its child the implementation in JSF use [here](https://github.com/nuxeo/nuxeo/blob/10.10/nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-web/src/main/java/org/nuxeo/elasticsearch/web/admin/ElasticSearchManager.java) initially i was going to implements the same thing on Management API side, but after i asked my self why i not use BAF and provide the query to get document and its child, i did it this way; i think it's better to reuse the same thing. but i have a doubt on the correctness of the query; direct child or traversal .... see [here](https://github.com/nuxeo/nuxeo/blob/00e74a9af58dde8c1d9066b85c5f76a52ceecb65/nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/work/IndexingWorker.java#L82) i think the query should be done on `ancestorId` anyway if the solution of providing the query instead of exposing directly is good for we will discuss about this query.

fyi: i took the choice to hide the implementation of how we index, this is the reason why in the return   method of my resource i just give the indexing id which behind the scene the commandId, same thing on my tests, its for the same reason that i decide to reuse baf instead of implementing the whole indexingCmd to index document and its children

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,170 @@
+     */
+    @POST
+    @Path("{repositoryName}/{documentId}/reindex")
+    public Response doIndexingOnDocument(@PathParam("repositoryName") String repositoryName,
```
IMHO, this is not needed, users can write a NXQL 🤔.

RSalem07 replied:
you means, we don't need to expose an endpoint to indexing by NXQL and Document ? if it's the case, with @ataillefer we decide to implement what is already done on JSF UI : https://nightly.nuxeo.com/nuxeo/nxadmin/default/default-domain@view_admin?tabIds=NUXEO_ADMIN%3AElasticSearch%3AESAdmin: `Re-indexing documents that match a NXQL query` and `Re-indexing a document and its children` :)

kevinleturc replied:
I meant that NXQL is enough.

RSalem07 replied:
I thought the same thing before adding this endpoint, but we don't have exactly the need, it's why we follow the jsf ui.

ataillefer replied:
I think we should ask the people who are needing the management API about this precise need.

ataillefer replied:
See https://github.com/nuxeo/nuxeo-management-rest-api/pull/9/commits/259ed646ef6aa0613a44bd94c93f2beeebafcb53#r306647668

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,170 @@
+     * @param indexingId the indexing id of the launched indexing
+     */
+    @GET
+    @Path("indexing/{indexingId}/status")
```
Could you follow what was introduced in `RenditionsObject`? Ie: no `/status` at the end of path and replace `indexingId` by `commandId`.

kevinleturc replied:
Following others comments, this endpoint could be `reindex/{commandId}`

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,170 @@
+        Map<String, BulkStatus.State> indexingStatus = new HashMap<>();
+        indexingStatus.put("status", bulkStatus.getState());
+
+        return Response.status(SC_OK).entity(indexingStatus).build();
```
Why not returning the `bulkStatus` directly?

RSalem07 replied:
yes, as i explained, my idea is to hide the implementation, from my point of view the client doesn't have to know that our implementation is BAF to index. It's why i decide to use a variable name as `indexingId` instead of `commandId`, and i didn't return the BulkStatus only the id. This way all is hided from outside,

kevinleturc replied:
Don't forget that users in this case are us, and by hiding an implementation detail, you also hide all metrics owned by the bulk status which are interested for the user. Last point, this is not consistent with what was done in `RenditionsObject`.

RSalem07 replied:
i was thinking that a customer user will use these endpoints, like it's done on JSF UI by admin users. I will modify the code ;)

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,170 @@
+
+    protected Response performIndexing(String repositoryName, String query) {
+        RepositoryService repositoryService = Framework.getService(RepositoryService.class);
+        if (repositoryService.getRepository(repositoryName) == null) {
```
This part is still not needed, this is handled by BAF. The only different thing I see is absence of HTTP status as BAF throws an `IllegalArgumentException`. Can you remove it?

RSalem07 replied:
yes i agree, my idea is to wrappe the http 500 into http 404, the exception message of BAF when there is an unknown repository contains the `toString` method of the command, and i can't based on it to translate it to an 404. plus that i think having an early control in this case saves us to launch the BAF, this is my point of view, But if you think it's really no need i will remove it and the test that control the 404 error.

kevinleturc replied:
A simpler solution is to improve the BAF code to replace this exception by a NuxeoException one with correct status.
Maybe this change won't need to be backported, TBD with the team.

The early control you're talking about is already implemented in the BAF service and is also early as we don't submit a command if repository doesn't exist.

RSalem07 replied:
i will create a jira ticket to add this improvements on BAF on 11.1, yes i agree i think no need to backport

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,170 @@
+     */
+    @POST
+    @Path("{repositoryName}/reindex")
+    public Response doIndexing(@PathParam("repositoryName") String repositoryName, @QueryParam("query") String query) {
```
What do you think to change `repositoryName` to a `QueryParam`? We have several places where we fallback on default repository if we don't explicitly provide a repository name. Note: This will be the case for at least 90% of environment.

In the same idea, we could do this change for other endpoints having such behavior (maybe with more code, don't know if ES services handles this default thing). But this will hide the _for all_ action 🤔 or we can handle such case with a wildcard, for instance `repositoryName = *`

RSalem07 replied:
yep, for the first proposition, i through about it, and i told my self perhaps if it's an error of the user (missing the query param in his url) then we don't need to launch an indexing on the default repository.  i think the ES service hold this behaviour, let me check i don't remember if it's only on automation.

good idea 👍 for the wildcard indexing (the all exiting repositories ;)) i will make some tests and i will tell you

RSalem07 replied:
if we don't provide a repository name, the default one will be taken ; [here](https://github.com/nuxeo/nuxeo/blob/564aa891032e778b757460f449792cf45cc6d862/nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/message/BulkCommand.java#L175) and [here](https://github.com/nuxeo/nuxeo/blob/564aa891032e778b757460f449792cf45cc6d862/nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/BulkServiceImpl.java#L98) i will modify the repositoryName to QueryParam instead of PathParam, but i don't know if we should add the Document and its children indexing, the idea is to transform the whole resources to QueryParam :
```
    @POST
    @Path("reindex")
    public Response doIndexing(@QueryParam("repositoryName") String repositoryName, @QueryParam("query") String query) {
        return performIndexing(repositoryName, query);
    }
```
 and i can't use the same thing on document, i will end a jaxrs error if i try to put 

```
   @POST
    @Path("reindex")
    public Response doIndexingOnDocument(@QueryParam("repositoryName") String repositoryName,
            @QueryParam("documentId") String documentId) {
        String query = String.format("Select * From Document where %s = '%s' or %s = '%s'", //
                NXQL.ECM_UUID, documentId, //
                NXQL.ECM_PARENTID, documentId);

        return performIndexing(repositoryName, query);
    }
```
@ataillefer @troger do you think we can do without "indexing document and its children", the user can use the query to index exactly what he need,

ataillefer replied:
Again, ask the people who will be using it. But I'm in favor of keeping it:
- it's cheap
- it's there in the JSF UI Admin Center
- it allow to reindex a subtree without reindexing the whole repository

If we decide to keep it, IMO, `documentId` should be a PathParam, which will solve your issue: 
```
@Path("reindex/{documentId}")
```

And you can still use an underlying NXQL query.

RSalem07 replied:
yep i see, i was in favour to put all as queryParam and to correct the issue i was going to make   `@Path("reindex/{documentId}")` but it's a little disturbing, because i see the repo as the high level, it's means we access to the repo and then the document. 
but to simplify and to keep it, i will do it as that ;)

kevinleturc replied:
Agree with @RSalem07 about the repository being a high level ressource, putting the `documentId` in the path whereas repository is not looks very weird to me. You'll end with:
```
curl -XPOST http://NUXEO/site/management/elasticsearch/reindex/00000000-0000-0000-0000-000000000000?repositoryName=default
```
Whereas NXQL can be independent from the repository on which we run it, the `documentId` is linked to repository and shouldn't be before/higher than the repository in the API.

Regarding the point above, only solution I see is:
```
@Path("reindex")
public Response reindex(@QueryParam("repositoryName") repositoryName, @QueryParam("query") query, @QueryParam("documentId") documentId) 
```
With this we need to handle the two cases depending on documentId presence.

RSalem07 replied:
yes i was thinking about doing it like that, as you mention, if the document is present the query will be ignored, if we decide to put document as more relevant for us. it's why i didn't do it yesterday :)

kevinleturc commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,170 @@
+                                       .path(ctx.getBasePath())
+                                       .path("/elasticsearch/indexing/{indexingId}/status")
+                                       .build(indexingId);
+        return Response.created(statusLocation).build();
```
Here it would be interesting to follow what was introduced in `RenditionsObject`. Ie: return the bulk status with an OK http status. Here the created location stuff can be error prone, `ctx.getBaseURL()` can return something which is not the public URI (for instance when nuxeo is behind a proxy) and this forces us to keep synchronize the path.

kevinleturc commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,233 @@
+        // Nothing indexed because of disable indexing flag
+        assertEquals(0, ess.query(new NxQueryBuilder(coreSession).nxql(GET_DOCUMENTS_QUERY)).totalSize());
+
+        // Start the ES indexing all document of the coreSession repository
```
```suggestion
        // Start the ES indexing of all document of the coreSession repository
```

ataillefer commented on nuxeo-management-rest-api/pom.xml:
```diff
@@ -45,6 +45,16 @@
+    <dependency>
+      <groupId>org.nuxeo.elasticsearch</groupId>
+      <artifactId>nuxeo-elasticsearch-automation</artifactId>
+    </dependency>
```
No more needed I guess

RSalem07 replied:
you right 👍 thx

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,131 @@
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Endpoint to administer Elasticsearch
```
Missing period

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,131 @@
+public class ElasticsearchObject extends DefaultObject {
+
+    /**
+     * Performs an ES indexing on documents of the given repository and that match the NXQL query.
```
the given NXQL query

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,131 @@
+            @PathParam("documentId") String documentId) {
+        String query = String.format("Select * From Document where %s = '%s' or %s = '%s'", //
+                NXQL.ECM_UUID, documentId, //
+                NXQL.ECM_PARENTID, documentId);
```
We want it recursive here, use `ancestorId`.

RSalem07 replied:
yep, i had a doubt about it, hence my question 👍

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    /**
+     * Performs an ES indexing on documents of the given repository and that match the NXQL query.
```
the given NXQL query

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,131 @@
+     *            of the given repository will be indexed
+     * @return the http response containing the location of the launched indexing
+     */
+
```
Extra blank line

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,131 @@
+            throw new NuxeoException(String.format("No such repository with name: %s", repositoryName), SC_NOT_FOUND);
+        }
+
+        String nxql = StringUtils.defaultIfBlank(query, "Select * From Document");
```
The default query could be in a constant.

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,202 @@
+        try (CloseableClientResponse clientResponse = httpClientRule.get(response.getLocation().getPath())) {
+            assertEquals(SC_OK, clientResponse.getStatus());
+            JsonNode jsonNode = mapper.readTree(clientResponse.getEntityInputStream());
+            assertEquals("COMPLETED", jsonNode.get("status").asText());
```
Do we have other interesting information in the status such as the number of successfully indexed documents?

RSalem07 replied:
yep, it `protected Long processed;`, i add the check on it + on query result.

RSalem07 replied:
as discuss the processed is available but on the status endpoint, as decided, a new endpoint will be created on BAF which allow retrieving the processed value. which means until that i don't see a way to control on the total documents treated by BAF :(

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,202 @@
+        String path = String.format(ES_INDEXING_PATH, coreSession.getRepositoryName())
+                + "?query=Select * From Document where dc:title like 'Title of my-file%'";
+
+        // Start the ES indexing document that match the nxql query (2 files)
```
of documents that match

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,202 @@
+        // Nothing indexed because of disable indexing flag
+        assertEquals(0, ess.query(new NxQueryBuilder(coreSession).nxql(GET_DOCUMENTS_QUERY)).totalSize());
+
+        // Start the ES indexing all document of the coreSession repository
```
of all documents

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,202 @@
+
+        // Start the ES indexing document that match the nxql query (2 files)
+        try (CloseableClientResponse response = httpClientRule.post(path, null)) {
+            verifyIndexingResponse(response, 2);
```
I wonder why you don't need to add the `initialDocCount` to the expected count as in the first test.

RSalem07 replied:
yes, it's because for each test i init/drop the indexes, and in this test i index only the document that match the query (2 files only) it's why i give only 2 as expected.

ataillefer commented on nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestElasticsearchObject.java:
```diff
@@ -0,0 +1,202 @@
+        DocumentModelList documents = coreSession.query("Select * from Document where dc:title = 'My Folder'");
+        String path = String.format("site/management/elasticsearch/%s/%s/reindex", coreSession.getRepositoryName(),
+                documents.get(0).getId());
+
```
To get the folder id you could also do:
```
coreSession.getDocument(new PathRef('/default-domain/workspaces/Folder")).getId()
```

RSalem07 replied:
done ;)

ataillefer commented:
Great job @RSalem07!

ataillefer commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -99,6 +100,44 @@ public Response getIndexingStatus(@PathParam("indexingId") String indexingId) {
+     */
+    @POST
+    @Path("{repositoryName}/flush")
+    public void doFlush(@PathParam("repositoryName") String repositoryName) {
```
I guess we also want to have the repo name as a `QueryParam`.

RSalem07 replied:
you right 👍

troger commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,127 @@
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Endpoint to administer Elasticsearch.
```
```suggestion
 * Endpoint to manage Elasticsearch.
```

RSalem07 replied:
:) i used manage and after that i changed it.

troger commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,127 @@
+                                                          .build());
+
+        BulkStatus bulkStatus = bulkService.getStatus(commandId);
+        return Response.status(SC_OK).entity(bulkStatus).build();
```
Just return the `bulkStatus` here, it will be marshalled automatically.

troger commented on nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/elasticsearch/ElasticsearchObject.java:
```diff
@@ -0,0 +1,127 @@
+    public static final String GET_ALL_DOCUMENTS_QUERY = "SELECT * from Document";
+
+    /**
+     * Performs an ES indexing on documents of the given repository and that match the given NXQL query.
```
I would have put the `repositoryName` as `@param` in the Javadoc to have a shorter summary sentence.
Something like:
```suggestion
     * Performs an ES indexing on documents matching the optional NXQL query.
     *
     * @param repositoryName the repository name on which we launch the indexing, cannot be {@code null} or {@code empty}
```

Same thing for all other methods 😬
